metals/src/main/scala/scala/meta/internal/decorations/ImplicitDecorationProvider.scala

```diff
@@ -0,0 +1,368 @@
+        range,
+        hover,
+        ThemableDecorationInstanceRenderOptions(
+          after = ThemableDecorationAttachmentRenderOptions(
+            text,
+            color = "grey",
+            fontStyle = "italic",
+            opacity = 0.7
+          )
+        )
+      )
+
+    val fullMessage = s"""|```scala
+                          |$hoverMessage
+                          |```""".stripMargin
+    // There is an issue hover for VS Code https://github.com/microsoft/vscode/issues/105302
+    val markup = new l.MarkupContent(
+      l.MarkupKind.MARKDOWN,
+      fullMessage
+    )
+
+    // sythetics get reported over multiple lines
+    val start = lspRange.getStart()
+    val end = lspRange.getEnd()
+    if (start.getLine() != end.getLine()) {
+      start.setLine(end.getLine())
+      start.setCharacter(0)
+    }
+    List(decoration(lspRange, decorationText, markup))
+  }
+
+  private def publishDecorations(
+      path: AbsolutePath,
+      textDocument: TextDocument
+  ): Unit = {
+    val decorations = for {
+      synthetic <- textDocument.synthetics
+      range <- synthetic.range.toIterable
+      edit = buffer.tokenEditDistance(path, textDocument.text)
+      decoration <- printSyntheticInfo(
+        textDocument,
+        synthetic,
+        toDecorationString(textDocument),
+        userConfig(),
+        simple = true
+      ).toIterable
+      hoverMessage <- printSyntheticInfo(
+        textDocument,
+        synthetic,
+        toHoverString(textDocument),
+        userConfig(),
+        simple = false
+      ).toIterable
+      lspRange <- edit.toRevisedStrict(range).toIterable
+      opt <- decorate(lspRange, decoration, hoverMessage)
+    } yield opt
+
+    val params =
+      new PublishDecorationsParams(path.toURI.toString(), decorations.toArray)
+
+    client.metalsPublishDecorations(params)
+  }
+
+  private def publishDiagnostics(
+      path: AbsolutePath,
+      textDocument: TextDocument
+  ) = {
+    val diags = for {
+      synthetic <- textDocument.synthetics
+      range <- synthetic.range.toIterable
+      edit = buffer.tokenEditDistance(path, textDocument.text)
+      fullMessage <- printSyntheticInfo(
+        textDocument,
+        synthetic,
+        toHoverString(textDocument),
+        userConfig(),
+        simple = false
+      ).toIterable
+      lspRange <- edit.toRevisedStrict(range).toIterable
+    } yield {
+
+      val diag = new l.Diagnostic(
+        lspRange,
+        fullMessage
+      )
+      diag.setSeverity(l.DiagnosticSeverity.Information)
+      diag
+    }
+    diagnostics.onPublishDiagnostics(path, diags, isReset = true)
+
+  }
+
+  private def publishImplicitInfo(
+      path: AbsolutePath,
+      textDocument: TextDocument
+  ): Future[Unit] =
+    Future {
+      if (clientConfig.isDecorationProvider()) {
+        publishDecorations(path, textDocument)
+      } else {
+        publishDiagnostics(path, textDocument)
+      }
+    }
+
+}
+
+object SemanticdbTreePrinter {
```

Comment:
I would probably move this to a separate file

Reply:
Moved!